docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerCmdExecFactory.java

```diff
@@ -0,0 +1,154 @@
+    public DefaultDockerCmdExecFactory(
+        DockerHttpClient dockerHttpClient,
+        ObjectMapper objectMapper
+    ) {
+        this.dockerHttpClient = dockerHttpClient;
+        this.objectMapper = objectMapper;
+    }
+
+    public DockerHttpClient getDockerHttpClient() {
+        return dockerHttpClient;
+    }
+
+    @Override
+    protected WebTarget getBaseResource() {
+        return new DefaultWebTarget();
+    }
+
+    @Override
+    public void close() throws IOException {
+        dockerHttpClient.close();
+    }
+
+    private class DefaultWebTarget implements WebTarget {
+
+        final ImmutableList<String> path;
+
+        final SetMultimap<String, String> queryParams;
+
+        DefaultWebTarget() {
+            this(
+                ImmutableList.of(),
+                MultimapBuilder.hashKeys().hashSetValues().build()
+            );
+        }
+
+        DefaultWebTarget(
+            ImmutableList<String> path,
+            SetMultimap<String, String> queryParams
+        ) {
+            this.path = path;
+            this.queryParams = queryParams;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("DefaultWebTarget{path=%s, queryParams=%s}", path, queryParams);
+        }
+
+        @Override
+        public InvocationBuilder request() {
+            String resource = StringUtils.join(path, "/");
+
+            if (!resource.startsWith("/")) {
+                resource = "/" + resource;
+            }
+
+            if (!queryParams.isEmpty()) {
+                Escaper urlFormParameterEscaper = UrlEscapers.urlFormParameterEscaper();
+                resource = queryParams.asMap().entrySet().stream()
+                    .flatMap(entry -> {
+                        return entry.getValue().stream().map(s -> {
+                            return entry.getKey() + "=" + urlFormParameterEscaper.escape(s);
+                        });
+                    })
+                    .collect(Collectors.joining("&", resource + "?", ""));
+            }
+
+            return new DefaultInvocationBuilder(
+                dockerHttpClient, objectMapper, resource
+            );
+        }
+
+        @Override
+        public DefaultWebTarget path(String... components) {
+            ImmutableList<String> newPath = ImmutableList.<String>builder()
+                .addAll(path)
+                .add(components)
+                .build();
+            return new DefaultWebTarget(newPath, queryParams);
+        }
+
+        @Override
+        public DefaultWebTarget resolveTemplate(String name, Object value) {
+            ImmutableList.Builder<String> newPath = ImmutableList.builder();
+            for (String component : path) {
+                component = component.replaceAll(
+                    "\\{" + name + "\\}",
+                    UrlEscapers.urlPathSegmentEscaper().escape(value.toString())
+                );
+                newPath.add(component);
+            }
```

Comment:
No Stream expression here?

Reply:
copied the code as-is, plus we need `ImmutableList`.